### PR TITLE
Flux new control models

### PR DESCRIPTION
**Pull request by andreasjansson**

### CODE REVIEW

File: flux/modules/image_embedders.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+        self.redux_dim = redux_dim
+        self.device = device if isinstance(device, torch.device) else torch.device(device)
+        self.dtype = dtype
+
+        with self.device:
+            self.redux_up = nn.Linear(redux_dim, txt_in_features * 3, dtype=dtype)
+            self.redux_down = nn.Linear(txt_in_features * 3, txt_in_features, dtype=dtype)
+
+            sd = load_sft(redux_path, device=str(device))
```
Comment (yorickvP) [inline flux/modules/image_embedders.py:104]: maybe wrap this in `if redux_path is not None`?

File: predict.py [hunk 2]
```diff
@@ -274,41 +324,35 @@ def compile_fp8(self):
...
-        )
-        img: torch.Tensor = transform(image)
-        return img[None, ...]
-
-    def get_mask(self, image: str):
-        if image is None:
-            return None
-        image = Image.open(image).convert("L")
+    # TODO(andreas): This is never used and will throw an error if
+    # used because aspect_ratio is not a valid argument to bf16_predict
```
Comment (yorickvP) [inline predict.py:328]: we could fix it with `width=1024, height=1024`, but it remains unused

File: fp8/flux_pipeline.py [hunk 3]
```diff
@@ -386,12 +386,12 @@ def into_bytes(self, x: torch.Tensor, jpeg_quality: int = 99) -> io.BytesIO:
         return im
 
     @torch.inference_mode()
-    def as_img_tensor(self, x: torch.Tensor) -> io.BytesIO:
-        """Converts the image tensor to bytes."""
```
Comment (daanelson) [inline fp8/flux_pipeline.py:390]: extremely nit but might be worth renaming this method to `as_pil_and_np_image` or something.

File: predict.py (hunk 2 above)
Comment (daanelson) [reply predict.py:328]: I say we just delete this; commented out code rots and this is already unused, we can pull this out of git history if we want to resurrect.

File: predict.py [hunk 4]
```diff
@@ -378,64 +440,62 @@ def handle_loras(
             self.bf16_lora = lora_weights
             self.bf16_lora_scale = lora_scale
 
-    def preprocess(self, aspect_ratio: str, megapixels: str = "1") -> Tuple[int, int]:
```
Comment (daanelson) [inline predict.py:381]: need to update this name change in the `HotswapPredictor`

File: predict.py [hunk 5]
```diff
@@ -446,9 +506,12 @@ def base_predict(
             dtype=torch.bfloat16,
             seed=seed,
         )
-        noise = x
         timesteps = get_schedule(
-            num_inference_steps, (x.shape[-1] * x.shape[-2]) // 4, shift=self.shift
+            num_inference_steps,
+            # TODO: this has changed in upstream flux to x.shape[1]
```
Comment (daanelson) [inline predict.py:511]: can probably delete this comment

**Review by daanelson — COMMENTED** [review]
generally great, need one update to `HotswapPredictor` and the rest is cosmetic. I should be able to push that in a bit.

**Review by daanelson — APPROVED** [review]

**Pull request merged**